File: languages/common/src/main/java/org/jetbrains/iren/settings/AppSettingsConfigurable.java
// Copyright 2000-2020 JetBrains s.r.o. and other contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.jetbrains.iren.settings;

import com.intellij.openapi.options.Configurable;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.iren.ngram.ModelBuilder;

import javax.swing.*;

/**
 * Provides controller functionality for application settings.
 */
public class AppSettingsConfigurable implements Configurable {

    private AppSettingsComponent mySettingsComponent;

    // A default constructor with no arguments is required because this implementation
    // is registered as an applicationConfigurable EP

    @Nls(capitalization = Nls.Capitalization.Title)
    @Override
    public String getDisplayName() {
        return "IRen";
    }

    @Override
    public JComponent getPreferredFocusedComponent() {
        return mySettingsComponent.getPreferredFocusedComponent();
    }

    @Nullable
    @Override
    public JComponent createComponent() {
        mySettingsComponent = new AppSettingsComponent();
        return mySettingsComponent.getPanel();
    }

    @Override
    public boolean isModified() {
        AppSettingsState settings = AppSettingsState.getInstance();
        try {
            boolean modified = mySettingsComponent.getMaxTrainingTime() != settings.maxTrainingTime;
            modified |= mySettingsComponent.getAutomaticTrainingStatus() != settings.automaticTraining;
            modified |= mySettingsComponent.getVocabularyCutOff() != settings.vocabularyCutOff;
            modified |= mySettingsComponent.getModelsLifetime() != settings.modelsLifetime;
            modified |= mySettingsComponent.getModelsLifetimeUnit() != settings.modelsLifetimeUnit;
            return modified;
        } catch (NumberFormatException e) {
            return false;
        }
    }

    @Override
    public void apply() {
        boolean needRetraining = needRetraining();
        AppSettingsState settings = AppSettingsState.getInstance();
        settings.automaticTraining = mySettingsComponent.getAutomaticTrainingStatus();
        settings.maxTrainingTime = mySettingsComponent.getMaxTrainingTime();
        settings.vocabularyCutOff = mySettingsComponent.getVocabularyCutOff();
        settings.modelsLifetime = mySettingsComponent.getModelsLifetime();
        settings.modelsLifetimeUnit = mySettingsComponent.getModelsLifetimeUnit();
        if (needRetraining) {
            ModelBuilder.trainModelsForAllProjectsInBackground();
        }
    }

    private boolean needRetraining() {
        AppSettingsState settings = AppSettingsState.getInstance();
//        Automatic training has been changed
        if (mySettingsComponent.getAutomaticTrainingStatus() != settings.automaticTraining)
            return mySettingsComponent.getAutomaticTrainingStatus();
//        Model training parameters have been changed
        return settings.automaticTraining &&
                (mySettingsComponent.getMaxTrainingTime() != settings.maxTrainingTime ||
                mySettingsComponent.getVocabularyCutOff() != settings.vocabularyCutOff);
    }

    @Override
    public void reset() {
        AppSettingsState settings = AppSettingsState.getInstance();
        mySettingsComponent.setMaxTrainingTime(settings.maxTrainingTime);
        mySettingsComponent.setAutomaticTrainingStatus(settings.automaticTraining);
        mySettingsComponent.setVocabularyCutOff(settings.vocabularyCutOff);
        mySettingsComponent.setModelsLifetime(settings.modelsLifetime);
        mySettingsComponent.setModelsLifetimeUnit(settings.modelsLifetimeUnit);
    }

    @Override
    public void disposeUIResources() {
        mySettingsComponent = null;
    }

}

File: plugin/src/main/kotlin/org/jetbrains/iren/settings/AppSettingsConfigurable.kt
package org.jetbrains.iren.settings

import com.intellij.openapi.options.Configurable
import org.jetbrains.annotations.Nls
import org.jetbrains.iren.settings.AppSettingsState.Companion.getInstance
import org.jetbrains.iren.training.ModelBuilder
import javax.swing.JComponent

/**
 * Provides controller functionality for application settings.
 */
class AppSettingsConfigurable : Configurable {
    private var mySettingsComponent: AppSettingsComponent? = null

    // A default constructor with no arguments is required because this implementation
    // is registered as an applicationConfigurable EP
    @Nls(capitalization = Nls.Capitalization.Title)
    override fun getDisplayName(): String {
        return "IRen"
    }

    override fun createComponent(): JComponent {
        mySettingsComponent = AppSettingsComponent()
        return mySettingsComponent!!.panel
    }

    override fun isModified(): Boolean {
        val settings = getInstance()
        return try {
            var modified = mySettingsComponent!!.maxTrainingTime != settings.maxTrainingTime
            modified = modified or (mySettingsComponent!!.automaticTrainingStatus != settings.automaticTraining)
            modified = modified or (mySettingsComponent!!.vocabularyCutOff != settings.vocabularyCutOff)
            modified = modified or (mySettingsComponent!!.modelsLifetime != settings.modelsLifetime)
            modified = modified or (mySettingsComponent!!.modelsLifetimeUnit != settings.modelsLifetimeUnit)
            modified
        } catch (e: NumberFormatException) {
            false
        }
    }

    override fun apply() {
        val needRetraining = needRetraining()
        val settings = getInstance()
        settings.automaticTraining = mySettingsComponent!!.automaticTrainingStatus
        settings.maxTrainingTime = mySettingsComponent!!.maxTrainingTime
        settings.vocabularyCutOff = mySettingsComponent!!.vocabularyCutOff
        settings.modelsLifetime = mySettingsComponent!!.modelsLifetime
        settings.modelsLifetimeUnit = mySettingsComponent!!.modelsLifetimeUnit
        if (needRetraining) {
            ModelBuilder.trainModelsForAllProjectsInBackground()
        }
    }

    private fun needRetraining(): Boolean {
        val settings = getInstance()
        //        Automatic training has been changed
        return if (mySettingsComponent!!.automaticTrainingStatus != settings.automaticTraining) mySettingsComponent!!.automaticTrainingStatus else settings.automaticTraining &&
                (mySettingsComponent!!.maxTrainingTime != settings.maxTrainingTime ||
                        mySettingsComponent!!.vocabularyCutOff != settings.vocabularyCutOff)
        //        Model training parameters have been changed
    }

    override fun reset() {
        val settings = getInstance()
        mySettingsComponent!!.maxTrainingTime = settings.maxTrainingTime
        mySettingsComponent!!.automaticTrainingStatus = settings.automaticTraining
        mySettingsComponent!!.vocabularyCutOff = settings.vocabularyCutOff
        mySettingsComponent!!.modelsLifetime = settings.modelsLifetime
        mySettingsComponent!!.modelsLifetimeUnit = settings.modelsLifetimeUnit
    }

    override fun disposeUIResources() {
        mySettingsComponent = null
    }
}